### DISCUSSION TITLE
User Defined Functions

### DISCUSSION THREAD

**Discussion by jimlloyd** [EvgSkv/logica#170] [Q&A] [answered]
Does Logica have direct support for User Defined Functions (UDFs)?

I imagine it might be straightforward to use Logical to generate SQL that could be used to implement a UDF with Standard SQL, but I wonder if there is something more direct.

I see Custom Aggregation Functions, but it's unclear how powerful that can be. The "as a function of a single positional argument" seems like it might quite restrictive compared to UDFs in SQL or Javascript that take multiple arguments, but as I am still early on the learning curve for Logica I could easily just not yet understand the full capability of custom aggregation functions.

**Comment by EvgSkv** [comment, marked as answer]
Yes!
And thanks to your comment I realized that tutorial should be improved to explain functions better.
There is no limit to how many arguments the function has, and arguments can be positional or named, just like with any other predicate.

Having one argument restriction is only for aggregation functions.

Here is an example of some functions:
```
@Engine("sqlite");

AreaOfRectangle(width, height) = width * height;
AreaOfTriangle(base:, height:) = base * height / 2;

Test("rectangle") = AreaOfRectangle(10, 20);
Test("triangle") = AreaOfTriangle(base: 10, height: 50);
```

Running it you'd get:

```
$ logica func.l run Test
+-----------+--------------+
| col0      | logica_value |
+-----------+--------------+
| rectangle | 200          |
| triangle  | 250          |
+-----------+--------------+
```

Please let me know if you have further questions.
